File: sdk/sdk-js/src/websocket.ts
/**
 * ReconnectingSocket — inline exponential-backoff WebSocket reconnect class.
 *
 * Connects to {wsBaseUrl}/sdk/ws/flags/{tenantId}, sends the SDK key as the
 * first text message immediately after `onopen` (first-message auth — the
 * backend `ws_router.py` expects a raw text frame, not JSON, within a
 * 10-second window; auth failure closes the socket with code 4001).
 *
 * On `onclose` (unless `close()` was called by the consumer), schedules a
 * reconnect with exponential backoff: `min(maxDelay, baseDelay * 2^attempt) + jitter`,
 * where `baseDelay=1000ms`, `maxDelay=30000ms`, and `jitter` is a random value
 * in `[0, 1000)ms`. The `attempt` counter resets to 0 on a successful `onopen`,
 * and the SDK-key handshake is repeated on every reconnect.
 *
 * `onMessage` receives the parsed JSON payload for every message (including
 * `{type:'ping'}` heartbeats — the caller decides what to ignore). Malformed
 * JSON is silently dropped (no exception escapes, `onMessage` is not called).
 *
 * No external dependencies — `reconnecting-websocket` (npm) is abandoned (2020).
 */
export class ReconnectingSocket {
  private ws: WebSocket | null = null
  private attempt = 0
  private readonly maxDelay = 30000
  private readonly baseDelay = 1000
  private closedByUser = false

  constructor(
    private url: string,
    private sdkKey: string,
    private onMessage: (data: any) => void,
  ) {
    this.connect()
  }

  private connect() {
    this.ws = new WebSocket(this.url)

    this.ws.onopen = () => {
      this.attempt = 0
      this.ws!.send(this.sdkKey)
    }

    this.ws.onmessage = (ev: MessageEvent) => {
      try {
        this.onMessage(JSON.parse(ev.data as string))
      } catch {
        // ignore malformed JSON
      }
    }

    this.ws.onclose = () => {
      if (!this.closedByUser) this.scheduleReconnect()
    }

    this.ws.onerror = () => {
      this.ws?.close()
    }
  }

  private scheduleReconnect() {
    const delay = Math.min(this.maxDelay, this.baseDelay * 2 ** this.attempt) + Math.random() * 1000
    this.attempt++
    setTimeout(() => this.connect(), delay)
  }

  /** Clean shutdown — prevents any further reconnect attempts. */
  close(): void {
    this.closedByUser = true
    this.ws?.close()
  }
}

File: sdk/sdk-python/src/backoffice_sdk/websocket.py
"""
Reconnecting WebSocket loop for backoffice_sdk live flag sync.

Connects to {ws_base_url}/sdk/ws/flags/{tenant_id}, performs first-message
auth (sends the raw sdk_key as the first text frame, matching
backend/app/domains/sdk/ws_router.py's protocol), and dispatches each
subsequent JSON message to an on_message callback. On disconnect, reconnects
with exponential backoff + jitter, capped at 30s, resetting the attempt
counter on successful connect.
"""
import asyncio
import json
import logging
import random
from typing import Callable

import websockets

logger = logging.getLogger(__name__)

BASE_DELAY = 1.0
MAX_DELAY = 30.0


def compute_backoff_delay(attempt: int) -> float:
    """Exponential backoff with jitter: min(30, 1 * 2^attempt) + random[0,1).
    Mirrors sdk-js ReconnectingSocket (Plan 08) and STATE.md v1.1 decision.
    """
    exp = min(MAX_DELAY, BASE_DELAY * (2 ** attempt))
    return exp + random.random()


async def ws_reconnect_loop(
    url: str,
    sdk_key: str,
    on_message: Callable[[dict], None],
    stop_event: "asyncio.Event | None" = None,
) -> None:
    """Connect to `url`, send `sdk_key` as the first text frame (first-message auth),
    then dispatch each subsequent JSON message to `on_message`. On disconnect,
    reconnect with exponential backoff + jitter; reset the attempt counter after
    a successful connection. Runs until cancelled or `stop_event` is set.
    """
    attempt = 0
    while stop_event is None or not stop_event.is_set():
        try:
            async with websockets.connect(url) as ws:
                await ws.send(sdk_key)
                attempt = 0  # reset on successful connect (post first-message send)
                async for raw in ws:
                    try:
                        msg = json.loads(raw)
                    except (TypeError, ValueError):
                        continue
                    on_message(msg)
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            logger.debug("SDK websocket disconnected: %s", exc)

        if stop_event is not None and stop_event.is_set():
            break

        delay = compute_backoff_delay(attempt)
        attempt += 1
        await asyncio.sleep(delay)
